src/Sylius/Component/User/Security/UserPbkdf2PasswordEncoder.php

```diff
@@ -0,0 +1,137 @@
+ *
+ * Providing a high level of Cryptographic security,
+ *  PBKDF2 is recommended by the National Institute of Standards and Technology (NIST).
+ *
+ * But also warrants a warning, using PBKDF2 (with a high number of iterations) slows down the process.
+ * PBKDF2 should be used with caution and care.
+ *
+ * @author Sebastiaan Stok <s.stok@rollerscapes.net>
+ * @author Andrew Johnson
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Michał Marcinkowski <michal.marcinkowski@lakion.com>
+ */
+class UserPbkdf2PasswordEncoder implements UserPasswordEncoderInterface
+{
+    const MAX_PASSWORD_LENGTH = 4096;
+
+    /**
+     * @var string
+     */
+    private $algorithm;
+
+    /**
+     * @var bool
+     */
+    private $encodeHashAsBase64;
+
+    /**
+     * @var int
+     */
+    private $iterations;
+
+    /**
+     * @var int
+     */
+    private $length;
+
+    /**
+     * @param string $algorithm          The digest algorithm to use
+     * @param bool   $encodeHashAsBase64 Whether to base64 encode the password hash
+     * @param int    $iterations         The number of iterations to use to stretch the password hash
+     * @param int    $length             Length of derived key to create
+     */
+    public function __construct($algorithm = 'sha512', $encodeHashAsBase64 = true, $iterations = 1000, $length = 40)
+    {
+        $this->algorithm = $algorithm;
+        $this->encodeHashAsBase64 = $encodeHashAsBase64;
+        $this->iterations = $iterations;
+        $this->length = $length;
+    }
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \LogicException when the algorithm is not supported
+     */
+    public function encode(UserInterface $user)
+    {
+        return $this->encodePassword($user->getPlainPassword(), $user->getSalt());
+    }
+
+    /**
+     * Encodes the raw password.
+     *
+     * @param string $raw  The password to encode
+     * @param string $salt The salt
+     *
+     * @return string The encoded password
+     *
+     * @throws \LogicException when the algorithm is not supported
+     */
+    protected function encodePassword($raw, $salt)
+    {
+        if ($this->isPasswordTooLong($raw)) {
+            throw new \InvalidArgumentException('Too long password.');
+        }
+
+        if (!in_array($this->algorithm, hash_algos(), true)) {
+            throw new \LogicException(sprintf('The algorithm "%s" is not supported.', $this->algorithm));
+        }
+
+        if (function_exists('hash_pbkdf2')) {
+            $digest = hash_pbkdf2($this->algorithm, $raw, $salt, $this->iterations, $this->length, true);
+        } else {
+            $digest = $this->hashPbkdf2($this->algorithm, $raw, $salt, $this->iterations, $this->length);
+        }
+
+        return $this->encodeHashAsBase64 ? base64_encode($digest) : bin2hex($digest);
+    }
+
+    private function hashPbkdf2($algorithm, $password, $salt, $iterations, $length = 0)
```

Comment:
Missing docblock and `@todo` to remove it in Sylius >=0.16 (when PHP5.5 will come).